src/main/java/com/linecorp/armeria/server/http/encoding/HttpEncodingService.java

```diff
@@ -0,0 +1,88 @@
+import com.linecorp.armeria.server.DecoratingService;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.http.HttpService;
+
+/**
+ * A {@link DecoratingService} that applies HTTP encoding (e.g., gzip) to an {@link HttpService}.
+ * HTTP encoding will be applied if the client specifies it, the response content type is a reasonable
+ * type to encode, and the response either has no fixed content length or the length is larger than 1KB.
+ */
+public class HttpEncodingService
+        extends DecoratingService<HttpRequest, HttpResponse, HttpRequest, HttpResponse> {
+
+    private static final Predicate<MediaType> DEFAULT_ENCODABLE_CONTENT_TYPE_PREDICATE =
+            contentType -> Stream.of(MediaType.ANY_TEXT_TYPE,
+                                     MediaType.APPLICATION_XML_UTF_8,
+                                     MediaType.JAVASCRIPT_UTF_8,
+                                     MediaType.JSON_UTF_8)
+                                 .anyMatch(contentType::is);
+
+    private static final int DEFAULT_MIN_BYTES_TO_FORCE_CHUNKED_AND_ENCODING = 1024;
+
+    private final Predicate<MediaType> encodableContentTypePredicate;
+    private final int minBytesToForceChunkedAndEncoding;
+
+    public HttpEncodingService(Service<? super HttpRequest, ? extends HttpResponse> delegate) {
+        this(delegate, DEFAULT_ENCODABLE_CONTENT_TYPE_PREDICATE,
+             DEFAULT_MIN_BYTES_TO_FORCE_CHUNKED_AND_ENCODING);
+    }
+
+    public HttpEncodingService(Service<? super HttpRequest, ? extends HttpResponse> delegate,
+                               Predicate<MediaType> encodableContentTypePredicate,
+                               int minBytesToForceChunkedAndEncoding) {
+        super(delegate);
+        this.encodableContentTypePredicate = requireNonNull(encodableContentTypePredicate,
+                                                            "encodableContentTypePredicate");
+        this.minBytesToForceChunkedAndEncoding = validateMinBytesToForceChunkedAndEncoding(
+                minBytesToForceChunkedAndEncoding);
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        HttpEncodingType encodingType = HttpEncoders.getWrapperForRequest(req);
+        HttpResponse delegateResponse = delegate().serve(ctx, req);
+        if (encodingType == null) {
+            return delegateResponse;
+        }
+        return new HttpEncodedResponse(
+                delegateResponse,
+                encodingType,
+                encodableContentTypePredicate,
+                minBytesToForceChunkedAndEncoding);
+    }
+
+    static int validateMinBytesToForceChunkedAndEncoding(int minBytesToForceChunkedAndEncoding) {
+        if (minBytesToForceChunkedAndEncoding <= 0) {
+            throw new IllegalArgumentException("minBytesToForceChunkedAndEncoding must be greater than 0.");
```

Comment:
I don't think there's a use case. On the flip side, since encoding implies chunking, it means that the empty response would be chunked as well for no good reason.